### PR TITLE
added products creation tab and general page composition

### PR DESCRIPTION
**Pull request by mmarganski**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,21 +1,37 @@
 {
-  "name": "cook",
+  "name": "todo-react",
```
Comment (gtarnopolski) [inline package.json]: cook I guess :D

File: package.json [hunk 2]
```diff
@@ -1,21 +1,37 @@
...
   "version": "0.1.0",
   "private": true,
   "dependencies": {
-    "@testing-library/jest-dom": "^5.11.4",
-    "@testing-library/react": "^11.1.0",
-    "@testing-library/user-event": "^12.1.10",
-    "react": "^17.0.2",
-    "react-dom": "^17.0.2",
+    "@testing-library/jest-dom": "5.11.4",
```
Comment (gtarnopolski) [inline package.json]: @tesing libraries are dev dependencies

File: src/components/Item.tsx [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import React from 'react'
+import styled from 'styled-components'
+
+export const Item = (props: {text: string}) => (
```
Comment (gtarnopolski) [inline src/components/Item.tsx]: we don't write types inline, create type ItemProps

File: src/components/Main.tsx [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+import { ProductsTab, RecipesTab, SearchTab, TabsBar } from './index'
+
+enum Tab {
+    Products,
+    Recipes,
+    Search
+}
+
+export const Main = () => {
+
```
Comment (gtarnopolski) [inline src/components/Main.tsx]: 5 lines under -> between consts we don't use enter

File: src/components/NameInput.tsx [hunk 5]
```diff
@@ -0,0 +1,50 @@
+import React, { useState } from 'react'
+import styled from 'styled-components'
+
+export const NameInput = (props: {onSubmittedInput(text: string): void}) => {
```
Comment (gtarnopolski) [inline src/components/NameInput.tsx]: .

File: src/components/ProductsTab.tsx [hunk 6]
```diff
@@ -0,0 +1,30 @@
+import React, { useState } from 'react'
+import styled from 'styled-components'
+import { ProductsView } from './ProductsView'
+import { NameInput } from './NameInput'
+
+export const ProductsTab = () => {
+
```
Comment (gtarnopolski) [inline src/components/ProductsTab.tsx]: .

File: src/components/ProductsView.tsx [hunk 7]
```diff
@@ -0,0 +1,26 @@
+import React from 'react'
+import styled from 'styled-components'
+import { Item } from './Item'
+
+export const ProductsView = (props: {products: Array<string>}) => {
+    window.localStorage.setItem('products', `${[...props.products]}`)
```
Comment (gtarnopolski) [inline src/components/ProductsView.tsx]:
```
JSON.Stringify({})
JSON.Parse()
```

**Review by gtarnopolski — CHANGES_REQUESTED** [review]

File: src/components/ProductsTab.tsx [hunk 8]
```diff
@@ -4,8 +4,7 @@ import { ProductsView } from './ProductsView'
 import { NameInput } from './NameInput'
 
 export const ProductsTab = () => {
-
-    const products = (window.localStorage.getItem('products') || '').split(',')
+    const products = JSON.parse(window.localStorage.getItem('products') || '[]')
```
Comment (gtarnopolski) [inline src/components/ProductsTab.tsx]: 👍

File: src/components/ProductsTab.tsx [hunk 9]
```diff
@@ -17,7 +16,7 @@ export const ProductsTab = () => {
     return (
         <Wrapper>
             <NameInput onSubmittedInput={addProduct}/>
-            <ProductsView products={currentState}/>
+            <ProductsView products={currentState} lastUpdate={''} selectable/>
```
Comment (gtarnopolski) [inline src/components/ProductsTab.tsx]: more than 2 props multiline :)

File: src/components/ProductsView.tsx [hunk 10]
```diff
@@ -1,25 +1,48 @@
...
-                    />
-                ))}
-        </Wrapper>
+        <ColumnWrapper>
+            <Header text={'Products'}/>
+            <RowWrapper>
+                {props.products
+                    .filter(product => product !== '')
+                    .map((product, index) => props.selectable
+                        ? (<Item
```
Comment (gtarnopolski) [inline src/components/ProductsView.tsx]:
```
                                 ? (
                                 <Item
                                      key={`${index}-${product}`}
                                      text={product}
                                    />
                            )  :  (
                                 <SelectWrapper
                                      key={`${index}-${product}`}
                                      text={product}
                                      lastUpdate={props.lastUpdate}
                        />
                     )
```

File: src/components/RecipesTab.tsx [hunk 11]
```diff
@@ -1,6 +1,48 @@
...
 
-export const RecipesTab = () => (
-    <p>Hello There</p>
-)
+export const RecipesTab = () => {
+    const products =  JSON.parse(window.localStorage.getItem('products') || '[]')
+    const recipes =  JSON.parse(window.localStorage.getItem('recipes') || '{}')
+    const [[currentRecipes, lastUpdate], setState] = useState([recipes, ''])
+
+    window.localStorage.removeItem('selected')
```
Comment (gtarnopolski) [inline src/components/RecipesTab.tsx]: this will be triggered each rerender (every time state changes)

File: src/components/RecipesView.tsx [hunk 12]
```diff
@@ -0,0 +1,38 @@
+import React from 'react'
+import styled from 'styled-components'
+import { Item } from './Item'
+import { Header } from './Header'
+
+type RecipesViewProps = {
+    recipes: any
```
Comment (gtarnopolski) [inline src/components/RecipesView.tsx]: no any

File: src/components/RecipesView.tsx [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+export const RecipesView = (props: RecipesViewProps) => {
+    window.localStorage.setItem('recipes', JSON.stringify(props.recipes))
```
Comment (gtarnopolski) [inline src/components/RecipesView.tsx]: aa

File: src/components/SelectWrapper.tsx [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+    if (props.lastUpdate !== lastRecipe) {
+        setState([false, props.lastUpdate])
+    }
+
+    const selectItem = () => {
+        const currentSelect = window.localStorage.getItem('selected')
+            ? JSON.parse(window.localStorage.getItem('selected') as string)
+            : []
+        const newSelect = isActive
+            ? [...currentSelect].filter(element => element !== props.text)
```
Comment (gtarnopolski) [inline src/components/SelectWrapper.tsx]: what is that [...currentSelect]

**Review by gtarnopolski — CHANGES_REQUESTED** [review]

File: src/App.js [hunk 15]
```diff
@@ -1,25 +1,8 @@
-import logo from './logo.svg';
-import './App.css';
+import './App.css'
```
Comment (jpudysz) [inline src/App.js]: import order

File: src/components/ClickableWrapper.tsx [hunk 16]
```diff
@@ -0,0 +1,24 @@
+import React from 'react'
+import styled from 'styled-components'
+
+type ClickableWrapperProps = {
+    children: React.ReactChild,
```
Comment (jpudysz) [inline src/components/ClickableWrapper.tsx]: no need to point `children` its already added underneath by `React.FunctionComponent<T>`

File: src/components/Header.tsx [hunk 17]
```diff
@@ -0,0 +1,21 @@
+import React from 'react'
+import styled from 'styled-components'
+
+type HeaderProps = {
+    text: string
+}
+
+export const Header: React.FunctionComponent<HeaderProps> = props => (
```
Comment (jpudysz) [inline src/components/Header.tsx]: we prefer destructiring

File: src/components/Item.tsx [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+type ItemProps = {
+    text: string,
+    isWrapped: boolean
+}
+
+type WrapperStyles = {
+    isWrapped: boolean
+}
+
+export const Item: React.FunctionComponent<ItemProps> = props => (
```
Comment (jpudysz) [inline src/components/Item.tsx]: aa

File: src/components/Main.tsx [hunk 19]
```diff
@@ -0,0 +1,51 @@
+import React, { useState } from 'react'
+import styled from 'styled-components'
+import { ProductsTab, RecipesTab, SearchTab, TabsBar } from './index'
```
Comment (jpudysz) [inline src/components/Main.tsx]: no `index` imports

File: src/components/Main.tsx [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+import { Tab } from '../types'
+
+export const Main: React.FunctionComponent = () => {
+    const [activeTab, setActiveTab] = useState(Tab.Products)
+
+    const onSelectTab= (tab: Tab) => {
```
Comment (jpudysz) [inline src/components/Main.tsx]: remove enter (group const) and add space after `onSelectTab`

File: src/components/Main.tsx [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+        switch (tab) {
+            case Tab.Products:
+                setActiveTab(Tab.Products)
+                break
+            case Tab.Recipes:
+                setActiveTab(Tab.Recipes)
+                break
+            case Tab.Search:
```
Comment (jpudysz) [inline src/components/Main.tsx:16]: can be shortened:
```
case Tab.Search:
default:
           setActiveTab(Tab.Search)
           break
```

File: src/components/Main.tsx [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+                break
+            default:
+                setActiveTab(Tab.Products)
+        }
+    }
+
+    const renderTabs = () => {
+        switch (activeTab) {
+            case Tab.Products:
+                return(<ProductsTab/>)
```
Comment (jpudysz) [inline src/components/Main.tsx]: add spacing:
```
return (
    <RecipesTab/>
)
```

File: src/components/NameInput.tsx [hunk 23]
```diff
@@ -0,0 +1,55 @@
+import React, { useState } from 'react'
+import styled from 'styled-components'
+
+type NameInputProps = {
+    onSubmittedInput(text: string): void
+}
+
+export const NameInput: React.FunctionComponent<NameInputProps> = props => {
```
Comment (jpudysz) [inline src/components/NameInput.tsx]: aa - destructirung

File: src/components/ProductsView.tsx [hunk 24]
```diff
@@ -0,0 +1,49 @@
...
+type ProductsViewProps = {
+    products: Array<string>,
+    isSelectable: boolean,
+    activeItems: Array<string>,
+    onSelect(text: string): void
+}
+
+export const ProductsView = (props: ProductsViewProps) => (
+    <ColumnWrapper>
+        <Header text={'Products'}/>
```
Comment (jpudysz) [inline src/components/ProductsView.tsx]: magic string + bad syntax for JSX's string

File: src/components/ProductsView.tsx [hunk 25]
```diff
@@ -0,0 +1,49 @@
...
+import { Header } from './Header'
+
+type ProductsViewProps = {
+    products: Array<string>,
+    isSelectable: boolean,
+    activeItems: Array<string>,
+    onSelect(text: string): void
+}
+
+export const ProductsView = (props: ProductsViewProps) => (
```
Comment (jpudysz) [inline src/components/ProductsView.tsx]: aa

File: src/components/RecipesTab.tsx [hunk 26]
```diff
@@ -0,0 +1,62 @@
...
+export const RecipesTab = () => {
+    const {getStorageRecipes, setStorageRecipes} = useLocalStorageRecipes()
+    const {getStorageProducts} = useLocalStorageProducts()
+    const [recipes, setRecipes] = useState(Object.keys(getStorageRecipes()))
+    const [activeItems, setActiveItems] = useState<Array<string>>([])
+
+    const addRecipe = (text: string) => {
+        if (activeItems.length > 0 && text !== '') {
+            setRecipes(prevRecipes => {
+                const storageRecipes =  getStorageRecipes()
```
Comment (jpudysz) [inline src/components/RecipesTab.tsx]: remove space

File: src/components/RecipesView.tsx [hunk 27]
```diff
@@ -0,0 +1,39 @@
...
+import { Item } from './Item'
+import { Header } from './Header'
+
+type RecipesViewProps= {
+    recipes: Array<string>
+}
+
+export const RecipesView: React.FunctionComponent<RecipesViewProps> = props => (
+    <ColumnWrapper>
+        <Header text="Recipes"/>
```
Comment (jpudysz) [inline src/components/RecipesView.tsx]: magic string

File: src/components/RecipesView.tsx [hunk 28]
```diff
@@ -0,0 +1,39 @@
...
+        <RowWrapper>
+            {props.recipes
+                .map((name, index) => (
+                    <Item
+                        key={`${index}-${name}`}
+                        text={name}
+                        isWrapped={false}
+                    />
+                )
+                )}
```
Comment (jpudysz) [inline src/components/RecipesView.tsx]: something went wrong here with the indentation

File: src/components/SearchTab.tsx [hunk 29]
```diff
@@ -0,0 +1,51 @@
...
+import styled from 'styled-components'
+import { ProductsView } from './ProductsView'
+import { RecipesView } from './RecipesView'
+import { useLocalStorageRecipes, useLocalStorageProducts } from '../hooks'
+
+export const SearchTab: React.FunctionComponent = () => {
+    const {getStorageRecipes} = useLocalStorageRecipes()
+    const {getStorageProducts} = useLocalStorageProducts()
+    const [recipes] = useState<Record<string, Array<string>>>(getStorageRecipes())
+    const [[activeProducts, activeRecipes], setActiveItems] = useState<Array<Array<string>>>([[], []])
```
Comment (jpudysz) [inline src/components/SearchTab.tsx]: for me it should be moved to two states

File: src/components/SelectWrapper.tsx [hunk 30]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+type SelectWrapperProps = {
+    text: string,
+    isActive: boolean,
+    onSelect(text: string): void
+}
+
+export const SelectWrapper: React.FunctionComponent<SelectWrapperProps> = props => (
+    <Wrapper onClick={() => props.onSelect(props.text)} isActive={props.isActive}>
```
Comment (jpudysz) [inline src/components/SelectWrapper.tsx]: one props per line

File: src/components/SelectWrapper.tsx [hunk 31]
```diff
@@ -0,0 +1,28 @@
...
+    isActive: boolean
+}
+
+type SelectWrapperProps = {
+    text: string,
+    isActive: boolean,
+    onSelect(text: string): void
+}
+
+export const SelectWrapper: React.FunctionComponent<SelectWrapperProps> = props => (
```
Comment (jpudysz) [inline src/components/SelectWrapper.tsx]: aa

File: src/components/TabsBar.tsx [hunk 32]
```diff
@@ -0,0 +1,30 @@
...
+import styled from 'styled-components'
+import { Header } from './Header'
+import { ClickableWrapper } from './ClickableWrapper'
+import { Tab } from '../types'
+
+type TabsBarProps = {
+    onSelect(tab: Tab): void
+}
+
+export const TabsBar: React.FunctionComponent<TabsBarProps> = props => (
```
Comment (jpudysz) [inline src/components/TabsBar.tsx]: aa

**Review by jpudysz — CHANGES_REQUESTED** [review]

File: src/components/ClickableWrapper.tsx [hunk 33]
```diff
@@ -0,0 +1,23 @@
+import React from 'react'
+import styled from 'styled-components'
+
+type ClickableWrapperProps = {
+    onSelect(): void
+}
+
+export const ClickableWrapper: React.FunctionComponent<ClickableWrapperProps> = ({onSelect, children}) => (
```
Comment (jpudysz) [inline src/components/ClickableWrapper.tsx]: you won't like it but you should add additional spaces between `{}` when you destructure :) (everywhere)

**Review by jpudysz — CHANGES_REQUESTED** [review]

File: src/components/ClickableWrapper.tsx (hunk 33 above)
Comment (jpudysz) [reply src/components/ClickableWrapper.tsx]: Apart of that, looks great, keep going!

File: src/components/RecipesTab.tsx [hunk 34]
```diff
@@ -6,16 +6,16 @@ import { RecipesView } from './RecipesView'
...
+    const { getStorageProducts } = useLocalStorageProducts()
     const [recipes, setRecipes] = useState(Object.keys(getStorageRecipes()))
     const [activeItems, setActiveItems] = useState<Array<string>>([])
 
     const addRecipe = (text: string) => {
         if (activeItems.length > 0 && text !== '') {
             setRecipes(prevRecipes => {
                 const storageRecipes = getStorageRecipes()
-                setStorageRecipes({...storageRecipes, [text]: activeItems})
+                setStorageRecipes({ ...storageRecipes, [text]: activeItems })
```
Comment (jpudysz) [inline src/components/RecipesTab.tsx]: split it into 3 lines:
```
setStorageRecipes({
     ...storageRecipes,
     [text]: activeItems   
})
```

**Review by jpudysz — CHANGES_REQUESTED** [review]

**Review by gtarnopolski — APPROVED** [review]

**Review by jpudysz — APPROVED** [review]

**Pull request merged**